### PR TITLE
T34551 remove system pages from history

### PR DESCRIPTION
**Pull request by dylanmccall**
https://phabricator.endlessm.com/T34551

### CODE REVIEW

**Review by manuq — APPROVED** [review]
The changes look good and relatively small, great! Let's try the APK and if all (manual) tests are passing let's merge.

**Review by manuq — CHANGES_REQUESTED** [review]
We found some regressions whlie testing it. Adding details to the ticket.

File: src/kolibri_android/main_activity/activity.py [hunk 1]
```diff
@@ -89,40 +81,53 @@ class MainActivity(BaseActivity):
...
             Runnable(self._on_start_with_usb),
             Runnable(self._on_loading_ready),
         )
 
     def on_activity_stopped(self, activity):
         super().on_activity_stopped(activity)
-        if self._kolibri_bus is not None:
-            self._kolibri_bus.transition("STOP")
+
+        if self._kolibri_bus.can_transition("IDLE"):
```
Comment (manuq) [inline src/kolibri_android/main_activity/activity.py]: Testing this in the onboarding (where Kolibri isn't running) I got the following Python traceback:
```
03-14 11:54:19.371  3871  3871 I python  : INFO:root:onActivityStopped
03-14 11:54:19.371  3871  3871 I python  : Traceback (most recent call last):
03-14 11:54:19.372  3871  3871 I python  :   File "jnius/jnius_proxy.pxi", line 50, in jnius.jnius.PythonJavaClass.invoke
03-14 11:54:19.372  3871  3871 I python  :   File "jnius/jnius_proxy.pxi", line 76, in jnius.jnius.PythonJavaClass._invoke
03-14 11:54:19.372  3871  3871 I python  :   File "/home/jenkins/workspace/kolibri-installer-android-pr/_cache/.local/share/python-for-android/build/python-installs/kolibri/x86_64/android/activity.py", line 172, in onActivityStopped
03-14 11:54:19.372  3871  3871 I python  :   File "/home/jenkins/workspace/kolibri-installer-android-pr/_cache/.local/share/python-for-android/build/python-installs/kolibri/x86_64/android/activity.py", line 88, in _callback
03-14 11:54:19.372  3871  3871 I python  :   File "/home/jenkins/workspace/kolibri-installer-android-pr/src/kolibri_android/main_activity/activity.py", line 99, in on_activity_stopped
03-14 11:54:19.372  3871  3871 I python  : AttributeError: 'NoneType' object has no attribute 'can_transition'
03-14 11:54:19.375  3871  3871 I python  : INFO:root:onActivitySaveInstanceState
```
So the check for `self._kolibri_bus is not None` has to be kept.

File: src/kolibri_android/main_activity/activity.py [hunk 2]
```diff
@@ -89,40 +81,53 @@ class MainActivity(BaseActivity):
...
+            logging.warning(
+                f"Kolibri is unable to stop because its state is '{self._kolibri_bus.state}"
+            )
 
     def on_activity_resumed(self, activity):
         super().on_activity_resumed(activity)
-        if self._kolibri_bus is not None:
-            self._kolibri_bus.transition("RUN")
+
+        if self._kolibri_bus.can_transition("START"):
```
Comment (manuq) [inline src/kolibri_android/main_activity/activity.py]: Same here, let's keep the check for None.

File: src/kolibri_android/main_activity/activity.py [hunk 3]
```diff
@@ -132,11 +137,20 @@ def on_activity_save_instance_state(self, activity, out_state_bundle):
...
+        self._last_url = None
+
+        if self._kolibri_bus.is_kolibri_url(current_url):
+            return urlparse(current_url)._replace(scheme="", netloc="").geturl()
+        else:
+            return None
+
     def run(self):
-        self.load_url("file:///android_asset/_load.html")
+        self.show_loading_screen()
```
Comment (manuq) [inline src/kolibri_android/main_activity/activity.py]: :+1:

File: src/kolibri_android/main_activity/kolibri_bus.py [hunk 4]
```diff
@@ -66,8 +69,8 @@ def SERVING(self, port):
         # if next_url is empty. This is resolved in Kolibri v0.16.0-alpha8:
         # <https://github.com/learningequality/kolibri/commit/0ed2ccdd5d613e96721f80bc03d8bc56ae7a0e7f>
         # TODO: Remove this workaround when we update Kolibri.
-        next_url = self.application.get_default_kolibri_path() or "/"
+        next_url = self.application.get_saved_kolibri_path() or "/"
         initialize_url = interface.get_initialize_url(next_url=next_url)
         logging.info(f"Using initialize URL '{initialize_url}'")
         start_url = urljoin(base_url, initialize_url)
-        self.application.load_url(start_url)
+        self.application.replace_url(start_url)
```
Comment (manuq) [inline src/kolibri_android/main_activity/kolibri_bus.py:76]: :+1:

File: src/kolibri_android/main_activity/activity.py [hunk 5]
```diff
@@ -89,40 +81,53 @@ class MainActivity(BaseActivity):
...
         super().on_activity_stopped(activity)
-        if self._kolibri_bus is not None:
-            self._kolibri_bus.transition("STOP")
+
+        if self._kolibri_bus.can_transition("IDLE"):
+            # With some versions of Android, the onSaveInstanceState hook will
+            # run after thsi one, so we need to keep track of the webview's
+            # URL before switching to the loading screen.
+            self._last_kolibri_path = self._get_current_kolibri_path()
+            self.show_loading_screen()
```
Comment (manuq) [inline src/kolibri_android/main_activity/activity.py]: I get the feeling that this only adds one more blinking screen when resuming. See the video I attached in the ticket. Now we have:
- The cached previously rendered page
- *** The "Welcome to Endless Key" (`assets/welcomeScreen/index.html`)
- The Kolibri bird
- The saved page (which may also have it's own loading state like animated cards)

*** This is the new one.

File: src/kolibri_android/main_activity/kolibri_bus.py [hunk 6]
```diff
@@ -49,6 +49,9 @@ def is_kolibri_url(self, url):
 
         return False
 
+    def can_transition(self, to_state: str) -> bool:
+        return (self.state, to_state) in self.transitions
```
Comment (manuq) [inline src/kolibri_android/main_activity/kolibri_bus.py:53]: I see you are adding type hints here. These are probably the first ones in this repo. Fine with me!

**Review by manuq — CHANGES_REQUESTED** [review]

**Comment by manuq** [comment]
@dylanmccall I'm bad with github suggestions UI, but here is a branch with 2 commits on top to fix what I'm seeing in testing: https://github.com/endlessm/kolibri-installer-android/pull/109

File: src/kolibri_android/main_activity/kolibri_bus.py (hunk 6 above)
Comment (dylanmccall) [reply src/kolibri_android/main_activity/kolibri_bus.py:53]: Oh oops, that was kind of by accident! Uh, doesn't hurt to leave them in I guess.

**Pull request merged**